include/deal.II/lac/trilinos_tpetra_vector.templates.h

```diff
@@ -643,8 +750,45 @@ namespace LinearAlgebra
 
     template <typename Number>
     void
-    Vector<Number>::compress(const VectorOperation::values /*operation*/)
-    {}
+    Vector<Number>::compress(const VectorOperation::values operation,
+                             const IndexSet               &locally_owned)
+    {
+      if (!compressed)
```

Comment:
I made calling `compress()` a crime on vectors with ghost elements in https://github.com/dealii/dealii/pull/16378 Perhaps you can add the same assertion here, and then you don't need the `if (!compressed)`.

Reply:
It seems like we have a clash of notation here:

Originally, I called a vector in this class compressed if and only if the vector can only read and write its locally owned elements. Only a non-compressed vector can read and write elements from other ranks.

As discussed above, I changed `compressed` to `has_ghost_elements` (where has_ghost_elements has the opposite meaning). And now it only makes sense to call `compress()` on vectors where `has_ghost_elements == true`...

Reply:
No, it's the other way around. Vectors with `has_ghost_values == true` are read-only. It does not make sense to call `compress()` on them.

Reply:
We looked into this together and found that the capabilities of the Epetra FEVector are no longer there.

So what we can no longer provide is a `reinit(local_set,MPI_Comm,omit_zeroing_entries)` as we need the `nonlocal_vector` to handle all incoming data as `SumIntoGlobalValues` will only work on locally owned global indices.

What we can do instead is actually use the `vector_writable` flag to construct a locally owned vector
as well as a nonlocal vector that stores all off-processor entries and communicates them during compress.

Sebastian is working on implementing this.
The question would now be what to do with the three argument reinit above. 
We see three options 
- Just don't implement it (could make porting code harder to understand)
- Implement it and throw an assertion (leads to completely undefined behaviour in release mode)
- Implement it and throw an exception 

I'm tending towards the last option. In the two 'implement' cases I would suggest the following text.
```
Due to differences between Tpetra and Epetra this capability is no longer supported.
To construct a locally owned vector use reinit(locally_owned,locally_relevant,mpi_communicator,/*writable_vector*/ true)
```

Reply:
About the `compressed` vs `has_ghost_elements` question: I'm ok with both approaches. Presumably, you want to use the former to indicate whether `compress()` has been called, and `has_ghost_elements` for what the name says? In the end, what I care about is the semantics that a vector that has ghost elements is read-only, and that consequently calling `compress()` on such a vector should either be a no-op or trigger an assertion.

As for what `reinit(local_set, communicator, omit_zeroing_entries)` should do: In the other vector classes, this creates a non-ghosted vector (which is consequently read-write) in which you can write into non-locally-owned elements (but not read from them) which are then communicated to their owners. A precondition for this function is that `local_set` is 1-to-1, i.e., the index sets provided on different processes are not overlapping.

Reply:
But that is what does not work or would amount to a lot of extra calculations and reinitializations
as the nonlocal vector itself has to be reinited.

If you look at the `SumIntoGlobalValue` [implementation](https://docs.trilinos.org/latest-release/packages/tpetra/doc/html/Tpetra__MultiVector__def_8hpp_source.html#l04150) you can see that it returns an error code
if the index is not part of the owned set.

So we could somehow infer the additional indices during add (i.e. whenever a new one comes up) 
and construct a vector out of that, or we could just take the knowledge that DoFHandler has 
and call reinit with the locally relevant set.

This functionality is what EpetraFEVector provided but it has no successor in Tpetra.

Reply:
Ah, that's a bummer. How does `parallel::distributed::Vector` do it? Perhaps we can steal the non-local stuff from there, steal the exchange mechanism from there, steal what `compress()` does from there, and in the end only implement the part where `p::d::Vector::compress()` writes into its own locally-owned part and here we just write into the Tpetra vector?

Reply:
To be honest I don't see how it does it.
It constructs a partitioner without ghost elements and checks in operator() whether or not the global index 
is part of either ghost indices or local indices. 

And I did not see a point at which the ghost indices of the partitioner are updated instead.

So I would hazard a guess that is also requires usage of the four input reinit with writable_vector.

As for the 'stealing' I'd take that from the Epetra class which has this as a fallback (maybe because FEVector wasn't always present).
There a Epetra Vector is constructed with a map out of the IndexSet relevant\local, i.e. only the ghost entries.
So we can split the input of `add` into local and nonlocal entries (with KokkosViews I think) and add to its corresponding Tpetra Vector. 
Then, compress simply does an Im/Export from the nonlocal to the local vector on the other processes.

Reply:
The only thing we really have to carefully think about and try out is whether we need an Importer or an Exporter. 
But I think Sebastian knows which one it is.

Reply:
Let me explain the related mechanisms of the `LinearAlgebra::distributed::Vector`:
- It has a two-argument `reinit(local_set, communicator)`, which creates a vector that has no access to ghost elements (neither read nor write). The internal implementation simply allocates a range for the local degrees of freedom and asserts that no access out of this bound is made. (This makes the vector not very useful for finite element computations in the regular way, but for many linear algebra operations this is enough, as this is what Tpetra does in their vector.)
- Then there is a three-argument `reinit(local_set, ghost_set, communicator)`, which is largely equivalent to the four-argument `reinit` discussed above. It allocates enough space for both the locally owned and ghost elements. In the default mode, the vector is writable, i.e., it does not expose read access to the ghost elements as it has not communicated them from the owner yet.
- The reason why the `writable` argument is not added in our own parallel vector is that the same class can switch between the two modes, `update_ghost_values()` makes the vector readable (but writable), whereas `zero_out_ghost_values()`, after which you cannot read ghosted elements (but you can write to it).

I entirely agree with the suggestion of @jpthiele and @kinnewig in the sense that for the writable vectors, we should `reinit` that vector by providing the owned elements as one index set and the ghost elements as a second index set. We can then set up a non-local vector and export/import the entries upon `compress()` into the owned vectors. The vector access functions then need to make sure if the index belongs in one or the other bucket, which is reasonable. The only other suitable way I see is to entirely manage the vector access in all assembly/affine constraints functions through our own `LinearAlgebra::distributed::Vector` and let `Tpetra` do the linear algebra with the locally owned part, i.e., let Tpetra never expose ghost entries (neither read nor write). I would not be opposed to that second option because it would make our internal interfaces leaner (less need to template all kinds of functions), but it likely needs significant work, as we would need to establish a better interface between the deal.II vectors and external linear algebra packages among others, and re-work most of the interfaces.

Let me comment a few more topics from above:
> As for the 'stealing' I'd take that from the Epetra class which has this as a fallback (maybe because FEVector wasn't always present).

This "fallback" was added later as an optimization, because `FEVector` is a lot slower then allocating the right non-local vector from the information in the DoFHandler (and because we realized in around 2014-2015 that the non-local vector would be the recommended way forward in the Trilinos/Tpetra world). In particular, we observed a very poor scaling in the parallel setting, because essentially all vector accesses would re-allocate entries, and because doing this on-demand is hard in either case.

> In the other vector classes, `reinit(local_set, communicator, omit_zeroing_entries)` creates a non-ghosted vector (which is consequently read-write) in which you can write into non-locally-owned elements (but not read from them) which are then communicated to their owners.

This is not what the `LA::d::Vector` does, and I would also recommend that we go away from such an interface. I would assume that more than 95% of use cases have the ability to predict which non-local entries will be added via the DoFHandler, so the default should be to provide this and not make the vector dynamically react on whatever entry a user might give it. There are such cases, but they are rare, and I do not think we should concentrate our effort on the unusual case.

As a final remark to the original post, I do not particularly like `operator=` between ghosted and non-ghosted vectors, because its effect is non-obvious to the user and it is brittle in the sense that a user needs a good understanding of the prerequisites. For `LA::d::Vector`, we needed many iterations to make the operator cover the relevant case of updating entries, while also doing what the user expects (a copy) for other cases. I would recommend to limit the use of `operator=` to the obvious ones of assignment/plain copy and not do things like ghost updates in future vector classes. But we can discuss that.

Reply:
Ah okay, well the history is what I was missing before. 
That makes sense and to me it felt like the most efficient way to handle the non local entries with the knowledge we get from the DoFHandler. 
If it is the more efficient way I wonder why it is not used more extensively in the tutorials. 
Is there a similar functionality in the PETSc wrappers? 

As for the three input case I just assumed it does the same as all other distributed classes and did not check. 
So good to know it behaves differently. 
It also makes sense to just keep one Vector from a memory standpoint and Sebastian and I shortly discussed this yesterday. 

So maybe we should do it the same way as `LA::d::Vector`
Since the old way does not work older Epetra codes have to be ported anyway, 
so they could be ported to a new principle.

The remaining question would then be whether we should change the PETSc wrappers as well so all distributed linear algebra works in the same fashion.

Reply:
@kronbichler, thank you for the detailed feedback. And the `FEVector` history is a very interesting side mark!

I just pushed my latest changes, where I added the `nonlocal_vector`. The `TpetraWrappers::Vector<Number>` current state works quite similarly to the old Epetra-based vector class. 

But at this point, it would be pretty easy to make the new Tpetra-based vector class more similar to the `LinearAlgebra::distributed::Vector` and use `vector_is_ghosted` in the same way as it is used in `LA::d::Vector`. This is mostly a design decision, therefore one of the principal developers has to say how we want to continue at this point.

Reply:
I am obviously a bit biased and would feel comfortable with the `LA::d::Vector` setup, but I do also think other options could work. I think we should have a common understanding here, so additional opinions would be appreciated.

Reply:
This is a long-standing disagreement between @kronbichler and me that we've had a friendly exchange about for probably the better part of a decade :-) 

My perspective is that it is useful to have a system in place that minimizes mistakes. Creating a vector so that it knows whether it can or cannot be written into is one such way, and it enables us to have assertions that check these semantics; it also avoids users to have to remember when they need to update ghost values. @kronbichler's perspective -- equally with merit -- is that we can come up with a system that does not require users to keep around two kinds of vectors, and that is perhaps more performant. 

I recognize that it is unfair to you to have to choose between the positions of two long-time developers of the library. My personal preference would be if we could get the semantics of the Tpetra wrappers as close as possible to those of the Epetra wrappers, principally because we would like to replace the latter by the former with the minimal amount of changes necessary -- ideally, our users will never notice that the underlying technology has changed. I recognize that @kronbichler has good reasons to argue for switching to his preferred approach as well.

Reply:
All right, 

I just pushed one last change, removing the now-unused `locally_(owned|relevant)_map`.
Now everything is ready from my side to be merged.

Reply:
I am also fine with having a minimal interface change between Epetra and Tpetra, and have the vector take either state (ghosted or non-ghosted) at the point it is constructed/initialized. I think the underlying mechanism with `nonlocal_vector` and manual import/export is uncontroversial, so the main mechanism is in place should we ever want to re-work the vector interfaces in a direction where a deep copy of the data would not be necessary to go from the read-only/ghosted state to the writable/non-ghosted state. I think there is consensus between @bangerth and me that starting to write into a ghosted vector leads to all kinds of problems because data can and will get inconsistent, and so one should help the user avoid this situation. The other things are not parts where I feel strongly.